### PR TITLE
Readd MAX-CUT rounding

### PR DESCRIPTION
**Pull request by blegat**
This MAX-CUT rounding is quite iconic, I would be sad to see it go. Follow up from https://github.com/jump-dev/JuMP.jl/pull/2914

### CODE REVIEW

File: docs/src/tutorials/conic/max_cut_sdp.jl [hunk 1]
```diff
@@ -40,13 +40,24 @@ function solve_max_cut_sdp(num_vertex, weights)
...
     @assert termination_status(model) == MOI.OPTIMAL
-    ## The paper uses some linear algebra to figure out which nodes belong to
-    ## which partition of the cut. We use a simpler way, and one that is more
-    ## robust to numerical error: comparing the rows, of which every element is
-    ## ±1.
-    opt_X = value.(X)
-    cut = map(1:num_vertex) do i
-        return isapprox(opt_X[1, :], opt_X[i, :]; atol = 1e-4) ? 1.0 : -1.0
+    ## Compute the Cholesky factorization of X, i.e., X = V^T V.
+    opt_X = LinearAlgebra.Hermitian(value.(X), :U)  # Tell Julia its PSD.
```
Comment (mlubin) [inline docs/src/tutorials/conic/max_cut_sdp.jl:44]: Typo: it's

Comment (blegat) [inline docs/src/tutorials/conic/max_cut_sdp.jl:44]:
```suggestion
    opt_X = LinearAlgebra.Hermitian(value.(X), :U)  # Tell Julia it's PSD.
```

**Comment by odow** [comment]
I didn't read the paper for specifics, but the problem is that it doesn't work as it extracts the wrong cut.

**Comment by odow** [comment]
Here's SCS@1.0
```Julia
julia> using JuMP

julia> import LinearAlgebra

julia> import SCS

julia> import Test

julia> weights = [
               0.0 1.0 5.0 0.0
               1.0 0.0 0.0 9.0
               5.0 0.0 0.0 2.0
               0.0 9.0 2.0 0.0
           ]
4×4 Matrix{Float64}:
 0.0  1.0  5.0  0.0
 1.0  0.0  0.0  9.0
 5.0  0.0  0.0  2.0
 0.0  9.0  2.0  0.0

julia> num_vertex = 4
4

julia> laplacian = LinearAlgebra.diagm(0 => weights * ones(num_vertex)) - weights
4×4 Matrix{Float64}:
  6.0  -1.0  -5.0   0.0
 -1.0  10.0   0.0  -9.0
 -5.0   0.0   7.0  -2.0
  0.0  -9.0  -2.0  11.0

julia> model = Model(SCS.Optimizer)
A JuMP Model
Feasibility problem with:
Variables: 0
Model mode: AUTOMATIC
CachingOptimizer state: EMPTY_OPTIMIZER
Solver name: SCS

julia> set_silent(model)

julia> @variable(
           model,
           X[i = 1:num_vertex, j = 1:num_vertex],
           PSD,
           start = (i == j ? 1.0 : 0.0),
       )
4×4 LinearAlgebra.Symmetric{VariableRef, Matrix{VariableRef}}:
 X[1,1]  X[1,2]  X[1,3]  X[1,4]
 X[1,2]  X[2,2]  X[2,3]  X[2,4]
 X[1,3]  X[2,3]  X[3,3]  X[3,4]
 X[1,4]  X[2,4]  X[3,4]  X[4,4]

julia> @objective(model, Max, 1 / 4 * LinearAlgebra.dot(laplacian, X))
1.5 X[1,1] - 0.5 X[1,2] - 2.5 X[1,3] + 2.5 X[2,2] - 4.5 X[2,4] + 1.75 X[3,3] - X[3,4] + 2.75 X[4,4]

julia> @constraint(model, LinearAlgebra.diag(X) .== 1)
4-element Vector{ConstraintRef{Model, MathOptInterface.ConstraintIndex{MathOptInterface.ScalarAffineFunction{Float64}, MathOptInterface.EqualTo{Float64}}, ScalarShape}}:
 X[1,1] = 1.0
 X[2,2] = 1.0
 X[3,3] = 1.0
 X[4,4] = 1.0

julia> optimize!(model)

julia> @assert termination_status(model) == MOI.OPTIMAL

julia> opt_X = value.(X)
4×4 Matrix{Float64}:
  1.0      -0.9999    -0.99999    1.00011
 -0.9999    1.0        0.999881  -0.999999
 -0.99999   0.999881   1.0       -1.00009
  1.00011  -0.999999  -1.00009    1.0

julia> import LinearAlgebra

julia> import Random

julia> opt_X = LinearAlgebra.Hermitian(value.(X), :U)
4×4 LinearAlgebra.Hermitian{Float64, Matrix{Float64}}:
  1.0      -0.9999    -0.99999    1.00011
 -0.9999    1.0        0.999881  -0.999999
 -0.99999   0.999881   1.0       -1.00009
  1.00011  -0.999999  -1.00009    1.0

julia> factorization = LinearAlgebra.cholesky(opt_X, Val(true); check = false)
LinearAlgebra.CholeskyPivoted{Float64, Matrix{Float64}}
U factor with rank 2:
4×4 LinearAlgebra.UpperTriangular{Float64, Matrix{Float64}}:
 1.0  -0.999999    -1.00009    1.00011
  ⋅    0.00124746  -0.168967   0.169824
  ⋅     ⋅          -0.028736  -0.99999
  ⋅     ⋅            ⋅         1.0
permutation:
4-element Vector{Int64}:
 4
 2
 3
 1

julia> V = (factorization.P * factorization.L)'
4×4 adjoint(::Matrix{Float64}) with eltype Float64:
  1.00011   -0.999999    -1.00009   1.0
  0.169824   0.00124746  -0.168967  0.0
 -0.99999    0.0         -0.028736  0.0
  1.0        0.0          0.0       0.0

julia> ## Normalize columns.
       for i in 1:num_vertex
           V[:, i] ./= LinearAlgebra.norm(V[:, i])
       end

julia> ## Generate random vector on unit sphere.
       Random.seed!(num_vertex)
MersenneTwister(4)

julia> r = rand(num_vertex)
4-element Vector{Float64}:
 0.680079235935741
 0.8744368931121158
 0.9240700869281602
 0.9293362188436958

julia> r /= LinearAlgebra.norm(r)
4-element Vector{Float64}:
 0.3963123329103651
 0.5095731596853206
 0.5384966229990681
 0.541565431623975

julia> ## Iterate over vertices, and assign each vertex to a side of cut.
       cut = ones(num_vertex)
4-element Vector{Float64}:
 1.0
 1.0
 1.0
 1.0

julia> for i in 1:num_vertex
           if LinearAlgebra.dot(r, V[:, i]) <= 0
               cut[i] = -1
           end
       end

julia> cut
4-element Vector{Float64}:
  1.0
 -1.0
 -1.0
  1.0
```

and here's SCS@1.1
```Julia
julia> using JuMP

julia> import LinearAlgebra

julia> import SCS

julia> import Test

julia> weights = [
               0.0 1.0 5.0 0.0
               1.0 0.0 0.0 9.0
               5.0 0.0 0.0 2.0
               0.0 9.0 2.0 0.0
           ]
4×4 Matrix{Float64}:
 0.0  1.0  5.0  0.0
 1.0  0.0  0.0  9.0
 5.0  0.0  0.0  2.0
 0.0  9.0  2.0  0.0

julia> num_vertex = 4
4

julia> laplacian = LinearAlgebra.diagm(0 => weights * ones(num_vertex)) - weights
4×4 Matrix{Float64}:
  6.0  -1.0  -5.0   0.0
 -1.0  10.0   0.0  -9.0
 -5.0   0.0   7.0  -2.0
  0.0  -9.0  -2.0  11.0

julia> model = Model(SCS.Optimizer)
A JuMP Model
Feasibility problem with:
Variables: 0
Model mode: AUTOMATIC
CachingOptimizer state: EMPTY_OPTIMIZER
Solver name: SCS

julia> set_silent(model)

julia> @variable(
           model,
           X[i = 1:num_vertex, j = 1:num_vertex],
           PSD,
           start = (i == j ? 1.0 : 0.0),
       )
4×4 LinearAlgebra.Symmetric{VariableRef, Matrix{VariableRef}}:
 X[1,1]  X[1,2]  X[1,3]  X[1,4]
 X[1,2]  X[2,2]  X[2,3]  X[2,4]
 X[1,3]  X[2,3]  X[3,3]  X[3,4]
 X[1,4]  X[2,4]  X[3,4]  X[4,4]

julia> @objective(model, Max, 1 / 4 * LinearAlgebra.dot(laplacian, X))
1.5 X[1,1] - 0.5 X[1,2] - 2.5 X[1,3] + 2.5 X[2,2] - 4.5 X[2,4] + 1.75 X[3,3] - X[3,4] + 2.75 X[4,4]

julia> @constraint(model, LinearAlgebra.diag(X) .== 1)
4-element Vector{ConstraintRef{Model, MathOptInterface.ConstraintIndex{MathOptInterface.ScalarAffineFunction{Float64}, MathOptInterface.EqualTo{Float64}}, ScalarShape}}:
 X[1,1] = 1.0
 X[2,2] = 1.0
 X[3,3] = 1.0
 X[4,4] = 1.0

julia> optimize!(model)

julia> @assert termination_status(model) == MOI.OPTIMAL

julia> opt_X = value.(X)
4×4 Matrix{Float64}:
  1.0  -1.0  -1.0   1.0
 -1.0   1.0   1.0  -1.0
 -1.0   1.0   1.0  -1.0
  1.0  -1.0  -1.0   1.0

julia> import LinearAlgebra

julia> import Random

julia> opt_X = LinearAlgebra.Hermitian(value.(X), :U)
4×4 LinearAlgebra.Hermitian{Float64, Matrix{Float64}}:
  1.0  -1.0  -1.0   1.0
 -1.0   1.0   1.0  -1.0
 -1.0   1.0   1.0  -1.0
  1.0  -1.0  -1.0   1.0

julia> factorization = LinearAlgebra.cholesky(opt_X, Val(true); check = false)
LinearAlgebra.CholeskyPivoted{Float64, Matrix{Float64}}
U factor with rank 1:
4×4 LinearAlgebra.UpperTriangular{Float64, Matrix{Float64}}:
 1.0  -1.0         -1.0   1.0
  ⋅   -5.01084e-7   1.0  -1.0
  ⋅     ⋅           1.0  -1.0
  ⋅     ⋅            ⋅    1.0
permutation:
4-element Vector{Int64}:
 1
 2
 3
 4

julia> V = (factorization.P * factorization.L)'
4×4 adjoint(::Matrix{Float64}) with eltype Float64:
 1.0  -1.0         -1.0   1.0
 0.0  -5.01084e-7   1.0  -1.0
 0.0   0.0          1.0  -1.0
 0.0   0.0          0.0   1.0

julia> ## Normalize columns.
       for i in 1:num_vertex
           V[:, i] ./= LinearAlgebra.norm(V[:, i])
       end

julia> ## Generate random vector on unit sphere.
       Random.seed!(num_vertex)
MersenneTwister(4)

julia> r = rand(num_vertex)
4-element Vector{Float64}:
 0.680079235935741
 0.8744368931121158
 0.9240700869281602
 0.9293362188436958

julia> r /= LinearAlgebra.norm(r)
4-element Vector{Float64}:
 0.3963123329103651
 0.5095731596853206
 0.5384966229990681
 0.541565431623975

julia> ## Iterate over vertices, and assign each vertex to a side of cut.
       cut = ones(num_vertex)
4-element Vector{Float64}:
 1.0
 1.0
 1.0
 1.0

julia> for i in 1:num_vertex
           if LinearAlgebra.dot(r, V[:, i]) <= 0
               cut[i] = -1
           end
       end

julia> cut
4-element Vector{Float64}:
  1.0
 -1.0
  1.0
 -1.0
```

**Comment by odow** [comment]
In the second case, the cholesky failed because the problem is rank deficient, but we skipped that check:
```Julia
factorization = LinearAlgebra.cholesky(opt_X, Val(true); check = false)
```

**Comment by blegat** [comment]
Thanks for the details, I'll take a look

**Comment by odow** [comment]
Coming back to this because I was like. Wait. The constraint is that X is PSD, so how can it return a rank deficient solution. 

And the key is this:
```Julia
julia> X
4×4 Matrix{Int64}:
  1  -1  -1   1
 -1   1   1  -1
 -1   1   1  -1
  1  -1  -1   1

julia> Y
4×4 Matrix{Float64}:
 1.0  0.0  0.0  0.0
 0.0  1.0  0.0  0.0
 0.0  0.0  1.0  0.0
 0.0  0.0  0.0  1.0

julia> isposdef(X)
false

julia> isposdef(X .+ 1e-15 * Y)
true
```
So in order to do the randomized algorithm, you'd need to first project the SCS matrix onto the PSD cone.

**Comment by blegat** [comment]
Yes, taking the cholesky decomposition of positive **semi**definite matrices is quite delicate, I usually use https://github.com/JuliaAlgebra/MultivariateMoments.jl/blob/fbe898e3fbfa40d7117db6f7c85c0e6f72625939/src/extract.jl#L84-L142 in SumOfSquares etc...

**Review by odow — APPROVED** [review]

**Pull request merged**